src/Renci.SshNet/SshCommand.cs

```diff
@@ -381,6 +384,107 @@ public string Execute(string commandText)
             return Execute();
         }
 
+#if NET6_0_OR_GREATER
+        /// <summary>
+        /// Executes command specified by <see cref="CommandText"/> property.
+        /// </summary>
+        /// <param name="token">The cancellation token.</param>
+        /// <returns>
+        /// Command execution result.
+        /// </returns>
+        /// <exception cref="SshConnectionException">Client is not connected.</exception>
+        /// <exception cref="SshOperationTimeoutException">Operation has timed out.</exception>
+        public async Task<string> ExecuteAsync(CancellationToken token)
+        {
+            // Prevent from executing BeginExecute before calling EndExecute
+            if (_asyncResult != null && !_asyncResult.EndCalled)
+            {
+                throw new InvalidOperationException("Asynchronous operation is already in progress.");
+            }
+
+            // Create new AsyncResult object
+            _asyncResult = new CommandAsyncResult
+                {
+                    AsyncWaitHandle = new ManualResetEvent(initialState: false),
+                    IsCompleted = false,
+                    AsyncState = null,
+                };
+
+            if (_channel is not null)
+            {
+                throw new SshException("Invalid operation.");
+            }
+
+            if (string.IsNullOrEmpty(CommandText))
+            {
+                throw new ArgumentException("CommandText property is empty.");
+            }
+
+            var outputStream = OutputStream;
+            if (outputStream is not null)
+            {
+                await outputStream.DisposeAsync().ConfigureAwait(false);
+                OutputStream = null;
+            }
+
+            var extendedOutputStream = ExtendedOutputStream;
+            if (extendedOutputStream is not null)
+            {
+                await extendedOutputStream.DisposeAsync().ConfigureAwait(false);
+                ExtendedOutputStream = null;
+            }
+
+            // Initialize output streams
+            OutputStream = new PipeStream();
+            ExtendedOutputStream = new PipeStream();
+
+            _result = null;
+            _error = null;
+            _callback = null;
+
+            _channel = CreateChannel();
+            await _channel.OpenAsync(token).ConfigureAwait(false);
+
+            _ = await _channel.SendExecRequestAsync(CommandText, token).ConfigureAwait(false);
+
+            lock (_endExecuteLock)
+            {
+                if (_asyncResult.EndCalled)
+                {
+                    throw new ArgumentException("EndExecute can only be called once for each asynchronous operation.");
+                }
+
+                // TODO I have no idea where it should be right now.
+                _inputStream?.Close();
+
+                // wait for operation to complete (or time out)
+                WaitOnHandle(_asyncResult.AsyncWaitHandle);
```

Comment:
Yeah.. before exposing async methods on SshCommand we should probably be at the point where its implementation is natively async (without AsyncResult and its wait handle), and the Begin/EndExecute methods just wrap the native methods with `TaskToAsyncResult`